sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -253,6 +256,9 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
   /**
    * Returns a copy of this node where `rule` has been recursively applied to it and all of its
    * children (pre-order). When `rule` does not apply to a given node it is left unchanged.
+   *
+   * Note that this function does not recurse into the children of a node that is being replaced.
```

Comment:
Do we need to explain this? This is so obvious that we can't recurse into the children of a node that is being replaced in a top-down transformation, or there is no way to define how the result will be.